// v-protocol/dist/vertica-types.js
"use strict";
/**
 * @license
 * Copyright (c) 2022-2024 Open Text.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.VerticaType = void 0;
var VerticaType;
(function (VerticaType) {
    VerticaType[VerticaType["Boolean"] = 5] = "Boolean";
    VerticaType[VerticaType["Integer"] = 6] = "Integer";
    VerticaType[VerticaType["Float"] = 7] = "Float";
    VerticaType[VerticaType["Char"] = 8] = "Char";
    VerticaType[VerticaType["Varchar"] = 9] = "Varchar";
    VerticaType[VerticaType["Varbinary"] = 17] = "Varbinary";
    VerticaType[VerticaType["LongVarchar"] = 115] = "LongVarchar";
    VerticaType[VerticaType["LongVarbinary"] = 116] = "LongVarbinary";
    VerticaType[VerticaType["Binary"] = 117] = "Binary";
    VerticaType[VerticaType["Numeric"] = 16] = "Numeric";
    VerticaType[VerticaType["Uuid"] = 20] = "Uuid";
    VerticaType[VerticaType["IntervalYear"] = 114] = "IntervalYear";
    VerticaType[VerticaType["IntervalYearToMonth"] = 114] = "IntervalYearToMonth";
    VerticaType[VerticaType["IntervalMonth"] = 114] = "IntervalMonth";
    VerticaType[VerticaType["IntervalDay"] = 14] = "IntervalDay";
    VerticaType[VerticaType["IntervalDayToHour"] = 14] = "IntervalDayToHour";
    VerticaType[VerticaType["IntervalDayToMinute"] = 14] = "IntervalDayToMinute";
    VerticaType[VerticaType["IntervalDayToSecond"] = 14] = "IntervalDayToSecond";
    VerticaType[VerticaType["IntervalHour"] = 14] = "IntervalHour";
    VerticaType[VerticaType["IntervalHourToMinute"] = 14] = "IntervalHourToMinute";
    VerticaType[VerticaType["IntervalHourToSecond"] = 14] = "IntervalHourToSecond";
    VerticaType[VerticaType["IntervalMinute"] = 14] = "IntervalMinute";
    VerticaType[VerticaType["IntervalMinuteToSecond"] = 14] = "IntervalMinuteToSecond";
    VerticaType[VerticaType["IntervalSecond"] = 14] = "IntervalSecond";
    VerticaType[VerticaType["Date"] = 10] = "Date";
    VerticaType[VerticaType["Time"] = 11] = "Time";
    VerticaType[VerticaType["TimeTz"] = 15] = "TimeTz";
    VerticaType[VerticaType["Timestamp"] = 12] = "Timestamp";
    VerticaType[VerticaType["TimestampTz"] = 13] = "TimestampTz";
    VerticaType[VerticaType["Row"] = 300] = "Row";
    VerticaType[VerticaType["Array"] = 301] = "Array";
    VerticaType[VerticaType["Map"] = 302] = "Map";
})(VerticaType = exports.VerticaType || (exports.VerticaType = {}));
